From 7464fabfe5351b6f488584d2c17b25116970c081 Mon Sep 17 00:00:00 2001
From: Md Sadre Alam <mdalam@qti.qualcomm.com>
Date: Thu, 26 Feb 2026 19:16:16 +0530
Subject: [PATCH] drivers: sdhci-msm: Guard ICE init code with
 CONFIG_MMC_CRYPTO

Fix boot failure when CONFIG_MMC_CRYPTO is disabled, by properly
guarding ICE initialization code. The ICE register writes were
executing unconditionally, causing hardware corruption and spurious
interrupts on systems without crypto support.

Move variable declarations and ICE initialization inside
CONFIG_MMC_CRYPTO guards to ensure code is only compiled when
crypto support is enabled.

Change-Id: Ib415bd68b1eefd8b46f706118dd642d21a7595e0
Signed-off-by: Md Sadre Alam <mdalam@qti.qualcomm.com>
---
 drivers/mmc/host/sdhci-msm.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/sdhci-msm.c b/drivers/mmc/host/sdhci-msm.c
index 904581ea6e78..bbffe3c94044 100644
--- a/drivers/mmc/host/sdhci-msm.c
+++ b/drivers/mmc/host/sdhci-msm.c
@@ -2115,8 +2115,6 @@ static int sdhci_msm_cqe_add_host(struct sdhci_host *host,
 	struct cqhci_host *cq_host;
 	bool dma64;
 	u32 cqcfg;
-	u32 config = 0x0;
-	u32 ice_cap = 0x0;
 	int ret;
 
 	/*
@@ -2171,7 +2169,11 @@ static int sdhci_msm_cqe_add_host(struct sdhci_host *host,
 	if (ret)
 		goto cleanup;
 
+#ifdef CONFIG_MMC_CRYPTO
 	/* Initalize the ICE for NON-CMDQ eMMC devices */
+	u32 config = 0x0;
+	u32 ice_cap = 0x0;
+
 	config = sdhci_readl(host, HC_VENDOR_SPECIFIC_FUNC4);
 	config &= ~DISABLE_CRYPTO;
 	sdhci_writel(host, config, HC_VENDOR_SPECIFIC_FUNC4);
@@ -2183,6 +2185,7 @@ static int sdhci_msm_cqe_add_host(struct sdhci_host *host,
 		cqhci_writel(cq_host, config, CQHCI_CFG);
 	}
 	sdhci_msm_ice_enable(msm_host);
+#endif
 
 	dev_info(&pdev->dev, "%s: CQE init: success\n",
 			mmc_hostname(host->mmc));
-- 
2.34.1

